File: modis_lst_vietnam.py
import ee
import geemap
import os
import datetime

# Authenticate với Earth Engine
ee.Initialize(project='ee-bonglantrungmuoi')

# Thiết lập thời gian
startDate = '2003-01-01'
endDate = '2025-01-01'

# Định nghĩa khu vực Việt Nam từ shapefile
viet_nam = ee.FeatureCollection("projects/ee-bonglantrungmuoi/assets/viet_nam")
viet_nam_geometry = viet_nam.geometry()

# Thiết lập thư mục đầu ra theo yêu cầu
output_dir = r"D:\HaiDang\MODISGapfilling\MODIS_LST_RAW"
os.makedirs(output_dir, exist_ok=True)

# Tạo thư mục con cho ảnh ngày và đêm
day_dir = os.path.join(output_dir, "Day")
night_dir = os.path.join(output_dir, "Night")
os.makedirs(day_dir, exist_ok=True)
os.makedirs(night_dir, exist_ok=True)

# Tải dữ liệu MODIS LST Terra và Aqua (raw)
modisTerra = ee.ImageCollection('MODIS/061/MOD11A1') \
    .filterDate(startDate, endDate) \
    .filterBounds(viet_nam) \
    .select(['LST_Day_1km', 'LST_Night_1km', 'QC_Day', 'QC_Night'])

modisAqua = ee.ImageCollection('MODIS/061/MYD11A1') \
    .filterDate(startDate, endDate) \
    .filterBounds(viet_nam) \
    .select(['LST_Day_1km', 'LST_Night_1km', 'QC_Day', 'QC_Night'])

# Chuyển các bộ dữ liệu thành danh sách ảnh
terra_list = modisTerra.toList(modisTerra.size())
aqua_list = modisAqua.toList(modisAqua.size())

print(f"Tổng số ảnh Terra: {modisTerra.size().getInfo()}")
print(f"Tổng số ảnh Aqua: {modisAqua.size().getInfo()}")

# Hàm tải xuống ảnh
def download_collection(collection_list, collection_name, total_images):
    for i in range(total_images):
        try:
            image = ee.Image(collection_list.get(i))
            image_date = ee.Date(image.get('system:time_start')).format('yyyy-MM-dd').getInfo()
            
            # Tạo ảnh riêng cho ban ngày và ban đêm
            day_image = image.select(['LST_Day_1km', 'QC_Day'])
            night_image = image.select(['LST_Night_1km', 'QC_Night'])
            
            # Tạo tên file xuất cho ảnh ngày
            day_filename = os.path.join(day_dir, f"{collection_name}_Day_{image_date}.tif")
            
            # Xuất ảnh ngày
            geemap.ee_export_image(
                ee_object=day_image,
                filename=day_filename,
                scale=1000,
                region=viet_nam_geometry,
                crs="EPSG:4326",
                file_per_band=False
            )
            
            # Tạo tên file xuất cho ảnh đêm
            night_filename = os.path.join(night_dir, f"{collection_name}_Night_{image_date}.tif")
            
            # Xuất ảnh đêm
            geemap.ee_export_image(
                ee_object=night_image,
                filename=night_filename,
                scale=1000,
                region=viet_nam_geometry,
                crs="EPSG:4326",
                file_per_band=False
            )
            
            print(f"Đã tải xuống {day_filename} và {night_filename}")
        except Exception as e:
            print(f"Lỗi khi tải ảnh {i} của {collection_name}: {e}")

# Tải xuống dữ liệu Terra và Aqua
print("Bắt đầu tải xuống dữ liệu Terra...")
download_collection(terra_list, "Terra", modisTerra.size().getInfo())

print("Bắt đầu tải xuống dữ liệu Aqua...")
download_collection(aqua_list, "Aqua", modisAqua.size().getInfo())

print(f"Tất cả ảnh đã được tải về thư mục: {output_dir}")
print(f"- Ảnh ngày: {day_dir}")
print(f"- Ảnh đêm: {night_dir}") 

File: MODIS_LST_Vietnam.js
// ===== 0. Thiết lập thời gian =====
var startDate = '2020-01-01';
var endDate = '2020-02-01';

var displayStart = ee.Date(startDate);
var displayEnd = displayStart.advance(2, 'day');

// ===== 1. Định nghĩa khu vực quan tâm từ shapefile Việt Nam =====
var viet_nam = ee.FeatureCollection("projects/ee-bonglantrungmuoi/assets/viet_nam");

// ===== 2. Tải dữ liệu MODIS LST =====
var modisTerra = ee.ImageCollection('MODIS/006/MOD11A1')
    .filterDate(startDate, endDate)
    .filterBounds(viet_nam)
    .select(['LST_Day_1km', 'LST_Night_1km', 'QC_Day', 'QC_Night']);

var modisAqua = ee.ImageCollection('MODIS/006/MYD11A1')
    .filterDate(startDate, endDate)
    .filterBounds(viet_nam)
    .select(['LST_Day_1km', 'LST_Night_1km', 'QC_Day', 'QC_Night']);

// ===== 3. Hàm lọc chất lượng =====
function filterQuality(image) {
    var qcDay = image.select('QC_Day');
    var qcNight = image.select('QC_Night');
    var qualityMaskDay = qcDay.bitwiseAnd(3).neq(3);
    var qualityMaskNight = qcNight.bitwiseAnd(3).neq(3);
    return image.addBands(image.select('LST_Day_1km').updateMask(qualityMaskDay).rename('filtered_LST_Day_1km'))
        .addBands(image.select('LST_Night_1km').updateMask(qualityMaskNight).rename('filtered_LST_Night_1km'));
}

var modisTerraFiltered = modisTerra.map(filterQuality);
var modisAquaFiltered = modisAqua.map(filterQuality);

// ===== 4. Kết hợp dữ liệu =====
var combined = modisTerraFiltered.merge(modisAquaFiltered);

// ===== 5. Xử lý riêng ban ngày và ban đêm (đổi từ K sang °C) =====
var dayData = combined.select('filtered_LST_Day_1km').map(function (image) {
    return image.multiply(0.02).subtract(273.15)
        .rename('LST_Day_1km_C')
        .set('system:time_start', image.get('system:time_start'));
});

var nightData = combined.select('filtered_LST_Night_1km').map(function (image) {
    return image.multiply(0.02).subtract(273.15)
        .rename('LST_Night_1km_C')
        .set('system:time_start', image.get('system:time_start'));
});

// ===== 6. Trung bình dài hạn =====
var meanDay = dayData.mean();
var meanNight = nightData.mean();

// ===== 7. Tính phần dư =====
var residualsDay = dayData.map(function (image) {
    return image.subtract(meanDay)
        .rename('residual_day')
        .set('system:time_start', image.get('system:time_start'));
});

var residualsNight = nightData.map(function (image) {
    return image.subtract(meanNight)
        .rename('residual_night')
        .set('system:time_start', image.get('system:time_start'));
});

// ===== 8. Làm mượt không gian =====
var smoothedResidualsDay = residualsDay.map(function (image) {
    return image.focal_mean({ radius: 3, kernelType: 'square', units: 'pixels' })
        .focal_mean({ radius: 3, kernelType: 'square', units: 'pixels' })
        .set('system:time_start', image.get('system:time_start'));
});

var smoothedResidualsNight = residualsNight.map(function (image) {
    return image.focal_mean({ radius: 3, kernelType: 'square', units: 'pixels' })
        .focal_mean({ radius: 3, kernelType: 'square', units: 'pixels' })
        .set('system:time_start', image.get('system:time_start'));
});

// ===== 9. Làm mượt thời gian =====
var temporalSmoothing = function (collection) {
    return collection.map(function (image) {
        var time = image.get('system:time_start');
        return ee.Algorithms.If(
            ee.Algorithms.IsEqual(time, null),
            image,
            function () {
                var date = ee.Date(time);
                var before = collection.filterDate(date.advance(-7, 'day'), date);
                var after = collection.filterDate(date, date.advance(7, 'day'));
                var window = before.merge(after).mean();
                return image.unmask(window)
                    .set('system:time_start', time);
            }()
        );
    });
};

var smoothedResidualsDayTemporal = temporalSmoothing(smoothedResidualsDay);
var smoothedResidualsNightTemporal = temporalSmoothing(smoothedResidualsNight);

// ===== 10. Tạo dữ liệu cuối cùng (clip theo shapefile Việt Nam) =====
var finalDay = smoothedResidualsDayTemporal.map(function (image) {
    return image.add(meanDay)
        .rename('final_LST_Day_1km_C')
        .clip(viet_nam)
        .set('system:time_start', image.get('system:time_start'));
});

var finalNight = smoothedResidualsNightTemporal.map(function (image) {
    return image.add(meanNight)
        .rename('final_LST_Night_1km_C')
        .clip(viet_nam)
        .set('system:time_start', image.get('system:time_start'));
});

// ===== 11. Hiển thị bản đồ =====
Map.centerObject(viet_nam, 6);

// Hiển thị ảnh LST final
var finalDayDisplay = finalDay.filterDate(displayStart, displayEnd);
var finalNightDisplay = finalNight.filterDate(displayStart, displayEnd);

Map.addLayer(finalDayDisplay, { min: -5, max: 35, palette: ['blue', 'green', 'yellow', 'red'] }, 'Final LST Day (Gap-filled, °C)');
Map.addLayer(finalNightDisplay, { min: -5, max: 35, palette: ['blue', 'green', 'yellow', 'red'] }, 'Final LST Night (Gap-filled, °C)');

// ===== 12. In thông tin kiểm tra =====
print('Final Day LST (°C):', finalDayDisplay);
print('Final Night LST (°C):', finalNightDisplay);

// ===== 13. Xuất ảnh đã gapfill cho 3 ngày cụ thể =====
// Chia Việt Nam thành 3 vùng theo latitude
var bounds = viet_nam.geometry().bounds();
var west = bounds.coordinates().get(0).get(0);
var south = bounds.coordinates().get(0).get(1);
var east = bounds.coordinates().get(0).get(2);
var north = bounds.coordinates().get(0).get(3);

// Tính toán các dải vĩ độ
var latRange = ee.Number(north).subtract(south);
var latStep = latRange.divide(3);

// Tạo danh sách các vùng
var regions = ee.List.sequence(0, 2).map(function (i) {
    var regionSouth = ee.Number(south).add(latStep.multiply(i));
    var regionNorth = regionSouth.add(latStep);
    var region = ee.Geometry.Rectangle([west, regionSouth, east, regionNorth]);
    return region;
});

// Ngày cần xuất
var exportDates = ['2020-01-01', '2020-01-15', '2020-02-01'];

// Xuất ảnh theo từng vùng và từng ngày
regions.evaluate(function (regionList) {
    regionList.forEach(function (region, regionIndex) {
        exportDates.forEach(function (date) {
            var start = ee.Date(date);
            var end = start.advance(1, 'day');

            // Lọc ảnh theo ngày
            var dayImage = finalDay.filterDate(start, end).mean();
            var nightImage = finalNight.filterDate(start, end).mean();

            // Xuất ảnh ban ngày
            Export.image.toDrive({
                image: dayImage,
                description: 'LST_Day_' + date + '_region' + (regionIndex + 1),
                folder: 'LST_Vietnam_Daily',
                region: region,
                scale: 1000,
                maxPixels: 1e13,
                crs: 'EPSG:4326'
            });

            // Xuất ảnh ban đêm
            Export.image.toDrive({
                image: nightImage,
                description: 'LST_Night_' + date + '_region' + (regionIndex + 1),
                folder: 'LST_Vietnam_Daily',
                region: region,
                scale: 1000,
                maxPixels: 1e13,
                crs: 'EPSG:4326'
            });
        });
    });
});